lib/ts/authUtils.ts

```diff
@@ -0,0 +1,426 @@
+            // then we do not want to include a link in the email.
+            for (const id of factorIds) {
+                // This util takes the tenant config into account (if it exists), then the MFA (static) config if it was initialized and set.
+                let validRes = await isValidFirstFactor(tenantId, id, userContext);
+
+                if (validRes.status === "TENANT_NOT_FOUND_ERROR") {
+                    if (session !== undefined) {
+                        throw new SessionError({
+                            type: SessionError.UNAUTHORISED,
+                            message: "Tenant not found",
+                        });
+                    } else {
+                        throw new Error("Tenant not found");
+                    }
+                } else if (validRes.status === "OK") {
+                    validFactorIds.push(id);
+                }
+            }
+
+            if (validFactorIds.length === 0) {
+                throw new SessionError({
+                    type: SessionError.UNAUTHORISED,
+                    message: "Session is required for secondary factors",
+                    payload: {
+                        clearTokens: false,
+                    },
+                });
+            }
+        } else {
+            // In this case the app will try to link the session user and the authenticating user after auth
+            // so we consider this to be an MFA factor setup (if the MFA recipe is initialized)
+            // and we check if factor setup is allowed accordingly
+            let caughtSetupFactorError;
+            if (mfaInstance !== undefined) {
+                const factorsAlreadySetUp = await MultiFactorAuth.getFactorsSetupForUser(
+                    session!.getUserId(userContext),
+                    userContext
+                );
+                // In all apis besides passwordless createCode, we know exactly which factor we are signing into, so in those cases,
+                // this is basically just checking if the single factor is allowed to be setup or not.
+                // For createCode (if the FE didn't pass the factor id exactly, which it should for MFA),
+                // we filter whatever is allowed. If any of them are allowed, createCode can happen.
+                // The filtered list can be used to select email templates. As an example:
+                // If the flowType for passwordless is USER_INPUT_CODE_AND_MAGIC_LINK and but only the otp-email factor is allowed to be set up
+                // then we do not want to include a link in the email.
+                for (const id of factorIds) {
+                    try {
+                        if (!factorsAlreadySetUp.includes(id)) {
+                            await MultiFactorAuth.assertAllowedToSetupFactorElseThrowInvalidClaimError(
+                                session!,
```

Comment:
we should take the authenticating user into account here and when we decide value of wantToLinkSessionUser